File: taurus-protect-sdk-python/taurus_protect/crypto/tpv1.py
"""TPV1-HMAC-SHA256 authentication for Taurus-PROTECT API."""

from __future__ import annotations

import hashlib
import hmac
import time
import uuid
from typing import Optional
from urllib.parse import urlparse


def _wipe_bytes(data: bytearray) -> None:
    """
    Securely wipe bytes from memory by overwriting with zeros.

    SECURITY NOTE: Python's garbage collector may retain copies of sensitive data
    in memory due to string interning, object copying, and reference counting.
    This wiping is best-effort and provides defense-in-depth, but cannot guarantee
    complete removal of sensitive data from process memory. For maximum security
    in production environments, consider using hardware security modules (HSMs)
    or dedicated secrets management solutions.
    """
    if data:
        for i in range(len(data)):
            data[i] = 0


class TPV1Auth:
    """
    TPV1-HMAC-SHA256 authentication handler.

    Generates Authorization headers for Taurus-PROTECT API requests.

    Example:
        auth = TPV1Auth("api-key", "hex-encoded-secret")
        header = auth.sign_request("POST", "api.example.com", "/v1/wallets", ...)
        # header = "TPV1-HMAC-SHA256 ApiKey=... Nonce=... Timestamp=... Signature=..."
    """

    def __init__(self, api_key: str, api_secret_hex: str) -> None:
        """
        Initialize TPV1 authentication.

        Args:
            api_key: The API key.
            api_secret_hex: The API secret as a hex-encoded string.

        Raises:
            ValueError: If api_key or api_secret_hex is empty.
        """
        self._closed = False  # Initialize early for __del__ safety
        self._secret = bytearray()

        if not api_key:
            raise ValueError("api_key cannot be empty")
        if not api_secret_hex:
            raise ValueError("api_secret cannot be empty")

        self.api_key = api_key
        # Store as bytearray for secure wiping
        self._secret = bytearray(bytes.fromhex(api_secret_hex))

    def close(self) -> None:
        """
        Securely wipe the API secret from memory.

        Note: Due to Python's garbage collector, string interning, and memory
        management, this is best-effort wiping. See _wipe_bytes() for details.
        """
        if not self._closed:
            _wipe_bytes(self._secret)
            self._closed = True

    def __del__(self) -> None:
        """Ensure secret is wiped on garbage collection."""
        self.close()

    def sign_request(
        self,
        method: str,
        host: str,
        path: str,
        query: Optional[str] = None,
        content_type: Optional[str] = None,
        body: Optional[str] = None,
    ) -> str:
        """
        Generate TPV1 Authorization header for a request.

        Args:
            method: HTTP method (GET, POST, etc.).
            host: Request host (e.g., "api.protect.taurushq.com").
            path: Request path (e.g., "/v1/wallets").
            query: Query string without leading "?" (optional).
            content_type: Content-Type header value (optional).
            body: Request body as string (optional).

        Returns:
            The Authorization header value.

        Raises:
            RuntimeError: If auth has been closed.
        """
        if self._closed:
            raise RuntimeError("TPV1Auth has been closed")

        nonce = str(uuid.uuid4())
        timestamp = int(time.time() * 1000)

        # Build message from non-empty parts
        parts = [
            "TPV1",
            self.api_key,
            nonce,
            str(timestamp),
            method.upper(),
            host,
            path,
        ]

        if query:
            parts.append(query)
        if content_type:
            parts.append(content_type)
        if body:
            parts.append(body)

        message = " ".join(parts)

        # Calculate HMAC-SHA256
        signature = hmac.new(
            bytes(self._secret),
            message.encode("utf-8"),
            hashlib.sha256,
        ).digest()

        import base64

        sig_b64 = base64.b64encode(signature).decode("utf-8")

        return (
            f"TPV1-HMAC-SHA256 ApiKey={self.api_key} "
            f"Nonce={nonce} Timestamp={timestamp} Signature={sig_b64}"
        )

    @staticmethod
    def parse_url(url: str) -> tuple[str, str, Optional[str]]:
        """
        Parse URL into host, path, and query components.

        Args:
            url: Full URL to parse.

        Returns:
            Tuple of (host, path, query).
        """
        parsed = urlparse(url)
        host = parsed.netloc
        path = parsed.path or "/"
        query = parsed.query if parsed.query else None
        return host, path, query


def calculate_base64_hmac(secret: bytes, data: str) -> str:
    """
    Calculate HMAC-SHA256 and return base64-encoded result.

    Args:
        secret: The secret key as bytes.
        data: The data to sign.

    Returns:
        Base64-encoded HMAC-SHA256 signature.
    """
    import base64

    signature = hmac.new(secret, data.encode("utf-8"), hashlib.sha256).digest()
    return base64.b64encode(signature).decode("utf-8")


def verify_base64_hmac(secret: bytes, data: str, expected_hmac: str) -> bool:
    """
    Verify HMAC-SHA256 signature using constant-time comparison.

    Args:
        secret: The secret key as bytes.
        data: The original data.
        expected_hmac: The expected base64-encoded HMAC.

    Returns:
        True if the HMAC matches.
    """
    computed = calculate_base64_hmac(secret, data)
    return hmac.compare_digest(computed, expected_hmac)

File: taurus-protect-sdk-typescript/src/crypto/tpv1.ts
/**
 * TPV1-HMAC-SHA256 authentication for Taurus-PROTECT API.
 *
 * Generates Authorization headers for API requests.
 */

import * as crypto from "crypto";

import { calculateBase64Hmac } from "./hashing";

/**
 * TPV1-HMAC-SHA256 authentication handler.
 *
 * Generates Authorization headers for Taurus-PROTECT API requests.
 *
 * @example
 * ```typescript
 * const auth = new TPV1Auth("api-key", "hex-encoded-secret");
 * const header = auth.signRequest("POST", "api.example.com", "/v1/wallets");
 * // header = "TPV1-HMAC-SHA256 ApiKey=... Nonce=... Timestamp=... Signature=..."
 *
 * // When done, securely wipe the secret
 * auth.close();
 * ```
 */
export class TPV1Auth {
  private readonly apiKey: string;
  private secret: Uint8Array;
  private closed: boolean = false;

  /**
   * Initialize TPV1 authentication.
   *
   * @param apiKey - The API key
   * @param apiSecretHex - The API secret as a hex-encoded string
   * @throws Error if apiKey or apiSecretHex is empty
   */
  constructor(apiKey: string, apiSecretHex: string) {
    if (!apiKey) {
      throw new Error("apiKey cannot be empty");
    }
    if (!apiSecretHex) {
      throw new Error("apiSecret cannot be empty");
    }

    this.apiKey = apiKey;
    this.secret = Buffer.from(apiSecretHex, "hex");
  }

  /**
   * Securely wipe the API secret from memory.
   *
   * After calling this method, the auth instance cannot be used to sign requests.
   *
   * **SECURITY NOTE - Memory Wiping Limitations in JavaScript:**
   * Due to JavaScript's garbage collector, copies of the secret may still exist
   * in memory even after this method is called. The V8 engine and other JavaScript
   * runtimes may create intermediate string copies, cache values, or defer garbage
   * collection indefinitely. This wiping is best-effort and provides defense-in-depth,
   * but cannot guarantee complete removal of sensitive data from memory.
   *
   * For environments with strict memory security requirements, consider:
   * - Using short-lived process isolation for sensitive operations
   * - Implementing HSM or secure enclave solutions for key management
   * - Avoiding long-lived processes that handle secrets
   */
  close(): void {
    if (!this.closed) {
      // Best-effort: Overwrite secret buffer with zeros
      // Note: This cannot prevent V8/GC from retaining copies elsewhere
      if (this.secret instanceof Buffer) {
        this.secret.fill(0);
      } else {
        for (let i = 0; i < this.secret.length; i++) {
          this.secret[i] = 0;
        }
      }
      this.closed = true;
    }
  }

  /**
   * Generate TPV1 Authorization header for a request.
   *
   * Message format: "TPV1 ApiKey Nonce Timestamp METHOD host path [query] [content-type] [body]"
   * Header format: "TPV1-HMAC-SHA256 ApiKey=X Nonce=Y Timestamp=Z Signature=W"
   *
   * @param method - HTTP method (GET, POST, etc.)
   * @param host - Request host (e.g., "api.protect.taurushq.com")
   * @param path - Request path (e.g., "/v1/wallets")
   * @param query - Query string without leading "?" (optional)
   * @param contentType - Content-Type header value (optional)
   * @param body - Request body as string (optional)
   * @returns The Authorization header value
   * @throws Error if auth has been closed
   *
   * @example
   * ```typescript
   * // Simple GET request
   * const header = auth.signRequest("GET", "api.protect.taurushq.com", "/v1/wallets");
   *
   * // POST with body
   * const header = auth.signRequest(
   *   "POST",
   *   "api.protect.taurushq.com",
   *   "/v1/wallets",
   *   undefined,
   *   "application/json",
   *   '{"name":"my-wallet"}'
   * );
   * ```
   */
  signRequest(
    method: string,
    host: string,
    path: string,
    query?: string,
    contentType?: string,
    body?: string
  ): string {
    if (this.closed) {
      throw new Error("TPV1Auth has been closed");
    }

    const nonce = this.generateNonce();
    const timestamp = Date.now();

    // Build message from non-empty parts
    // Format: "TPV1 ApiKey Nonce Timestamp METHOD host path [query] [content-type] [body]"
    const parts: string[] = [
      "TPV1",
      this.apiKey,
      nonce,
      timestamp.toString(),
      method.toUpperCase(),
      host,
      path,
    ];

    // Add optional parts only if they have values
    if (query) {
      parts.push(query);
    }
    if (contentType) {
      parts.push(contentType);
    }
    if (body) {
      parts.push(body);
    }

    const message = parts.join(" ");

    // Calculate HMAC-SHA256 signature
    const signature = calculateBase64Hmac(this.secret, message);

    // Format: "TPV1-HMAC-SHA256 ApiKey=X Nonce=Y Timestamp=Z Signature=W"
    return `TPV1-HMAC-SHA256 ApiKey=${this.apiKey} Nonce=${nonce} Timestamp=${timestamp} Signature=${signature}`;
  }

  /**
   * Generate a unique nonce for request signing.
   *
   * Uses Node.js built-in crypto.randomUUID() (available in Node.js 14.17+).
   *
   * @returns UUID v4 string
   */
  private generateNonce(): string {
    return crypto.randomUUID();
  }

  /**
   * Parse URL into host, path, and query components.
   *
   * @param url - Full URL to parse
   * @returns Object containing host, path, and optional query
   *
   * @example
   * ```typescript
   * const { host, path, query } = TPV1Auth.parseUrl("https://api.example.com/v1/wallets?limit=10");
   * // host = "api.example.com"
   * // path = "/v1/wallets"
   * // query = "limit=10"
   * ```
   */
  static parseUrl(url: string): { host: string; path: string; query?: string } {
    const parsed = new URL(url);
    return {
      host: parsed.host,
      path: parsed.pathname || "/",
      query: parsed.search ? parsed.search.substring(1) : undefined,
    };
  }
}

/**
 * Calculate the TPV1 signed header value.
 *
 * This is a standalone function for use cases where you don't want to maintain
 * a TPV1Auth instance.
 *
 * @param apiKey - The API key
 * @param apiSecret - The API secret as bytes
 * @param nonce - Unique nonce for this request
 * @param timestamp - Request timestamp in milliseconds
 * @param method - HTTP method
 * @param host - Request host
 * @param path - Request path
 * @param query - Query string (optional)
 * @param contentType - Content-Type header (optional)
 * @param body - Request body (optional)
 * @returns The Authorization header value
 */
export function calculateSignedHeader(
  apiKey: string,
  apiSecret: Uint8Array,
  nonce: string,
  timestamp: number,
  method: string,
  host: string,
  path: string,
  query?: string,
  contentType?: string,
  body?: string
): string {
  // Build message from non-empty parts
  const parts: string[] = [
    "TPV1",
    apiKey,
    nonce,
    timestamp.toString(),
    method.toUpperCase(),
    host,
    path,
  ];

  if (query) {
    parts.push(query);
  }
  if (contentType) {
    parts.push(contentType);
  }
  if (body) {
    parts.push(body);
  }

  const message = parts.join(" ");
  const signature = calculateBase64Hmac(apiSecret, message);

  return `TPV1-HMAC-SHA256 ApiKey=${apiKey} Nonce=${nonce} Timestamp=${timestamp} Signature=${signature}`;
}
